include/mppic/parameters_handler.hpp

```diff
@@ -0,0 +1,173 @@
+  {
+    node_ = parent;
+    auto node = node_.lock();
+    logger_ = node->get_logger();
+  }
+
+  void start()
+  {
+    auto node = node_.lock();
+    on_set_param_handler_ = node->add_on_set_parameters_callback(
+      std::bind(
+        &ParametersHandler::dynamicParametersCallback, this, std::placeholders::_1));
+  }
+
+  rcl_interfaces::msg::SetParametersResult dynamicParametersCallback(
+    std::vector<rclcpp::Parameter> parameters)
+  {
+    rcl_interfaces::msg::SetParametersResult result;
+
+    std::lock_guard<std::mutex> lock(parameters_change_mutex_);
+
+    for (auto & pre_cb : pre_callbacks_) {
+      pre_cb();
+    }
+
+    for (auto & param : parameters) {
+      const std::string & param_name = param.get_name();
+
+      if (auto callback = get_param_callbacks_.find(param_name);
+        callback != get_param_callbacks_.end())
+      {
+        RCLCPP_INFO(logger_, "Parameter %s found", param_name.c_str());
+        callback->second(param);
+      } else {
+        RCLCPP_WARN(logger_, "Parameter %s not found", param_name.c_str());
+      }
+    }
+
+    for (auto & post_cb : post_callbacks_) {
+      post_cb();
+    }
+
+    result.successful = true;
+    return result;
+  }
+
+  std::mutex * getLock()
+  {
+    return &parameters_change_mutex_;
+  }
+
+
+  template<typename T>
+  void addPostCallback(T && callback)
+  {
+    post_callbacks_.push_back(std::move(callback));
+  }
+
+  template<typename T>
+  void addPreCallback(T && callback)
+  {
+    pre_callbacks_.push_back(std::move(callback));
+  }
+
+  auto getParamGetter(std::string ns)
+  {
+    return [this, ns = std::move(ns)](auto & setting, std::string name, auto default_value,
+             ParameterType param_type = ParameterType::Dynamic) {
+             getParam(
+               setting, ns == "" ? name : ns + "." + name, std::move(
+                 default_value), param_type);
+           };
+  }
+
+  template<typename SettingT, typename ParamT>
+  void getParam(
+    SettingT & setting, std::string name, ParamT default_value,
+    ParameterType param_type = ParameterType::Dynamic)
+  {
+    auto node = node_.lock();
+
+    nav2_util::declare_parameter_if_not_declared(
+      node, name, rclcpp::ParameterValue(default_value));
+
+    setStaticParam<ParamT>(node, setting, name);
+
+    if (param_type == ParameterType::Dynamic) {
+      setDynamicParamCallback(setting, name);
+    }
+  }
+
+  template<typename ParamT, typename SettingT, typename NodeT>
+  void setStaticParam(NodeT node, SettingT & setting, std::string name)
+  {
+    ParamT param_in;
+    node->get_parameter(name, param_in);
+    setting = static_cast<SettingT>(param_in);
+  }
+
+  template<typename T>
+  void setDynamicParamCallback(
+    T & setting, std::string name)
+  {
+
+    if (auto it = get_param_callbacks_.find(name);
+      it != get_param_callbacks_.end())
+    {
+      return;
+    }
+
+    get_param_callbacks_[name] = [this, &setting, name](const rclcpp::Parameter & param) {
```

Comment:
The Parameter class contains the type infromation `[get_type](https://docs.ros2.org/foxy/api/rclcpp/classrclcpp_1_1Parameter.html#ac845c383400a1264f168cca55daab110)` or even [get_value](https://docs.ros2.org/foxy/api/rclcpp/classrclcpp_1_1Parameter.html#a9042da271855312c63c3db86162c9b2a) is going to return the auto type

Reply:
that might make this a little more clean without the need of more complex templating standard library functions to figure out types

Reply:
Nothing about what you did here is wrong, but it can be cleaned up with Parameter and ParameterValue information that is provided to you

Reply:
get_type isn't constexpr function (so i cant know result of this at compile time). So i cant use result of get_type in template argument of get_value, but i do know at compile time optimizer parameters, so i can use it that way

Reply:
What I'm saying is that you don't need to use any of that `is_same_v` stuff. You can just do `if param.get_type() == PARAMETER_BOOL` to check the types of the parameters. The argument itself contains enough information to deduce the type

Reply:
Yeah, i got it, but as far as i know this deduction will be in runtime. Am i wrong? 
Here i wanted compiler to generate branch depending on which type has setting. 
The same stuff actually, but at compile time. I dont think it will be cleaner if change this to ros style deduction.  Few lines of code would differ

Reply:
But why is branching a more optimal method for a set of callbacks that really shouldn't be executed frequently during runtime? I would think readability of using methods provided by objects we have available to us is more clear / readable and not measurably different in performance. I'm not saying what you have is technically wrong, but generally speaking I try to avoid semantics like that if at all possible due to the non-native semantics they give (e.g. its difficult to read that an immediately know what each line means, its not overly complex, but its also not dead simple). I think of using those tools as a measure of last resort when I don't have easy access to type information and I must do it that way to deduce types for operations. But if you know what it is outright, isn't it better to use that?